core/src/main/java/org/elasticsearch/index/query/RandomSampleQuery.java

```diff
@@ -0,0 +1,318 @@
+                            max = Math.min(max, maxDoc);
+                            FakeScorer scorer = new FakeScorer();
+                            scorer.score = score;
+                            collector.setScorer(scorer);
+
+                            int current = min;
+                            while (current < max) {
+                                int gap = getGap(rng);
+                                current = current + gap;
+                                if (current >= maxDoc) {
+                                    return DocIdSetIterator.NO_MORE_DOCS;
+                                }
+                                scorer.doc = current;
+                                if (acceptDocs == null || acceptDocs.get(current)) {
+                                    collector.collect(current);
+                                }
+                            }
+                            return max == maxDoc ? DocIdSetIterator.NO_MORE_DOCS : max;
+                        }
+                        @Override
+                        public long cost() {
+                            return maxDoc;
+                        }
+                    };
+                }
+
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "RandomSample[p=" + this.p + "](*:*)";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        RandomSampleQuery other = (RandomSampleQuery)o;
+        return Objects.equals(this.p, other.p) &&
+            Objects.equals(this.seed, other.seed);
+    }
+
+    @Override
+    public int hashCode() {
+        return classHash();
+    }
+
+    /**
+     * A DocIDSetIter that samples on each document.  Empirically, this tends to
+     * be more accurate when the segment is small.  It may also be faster
+     * when the probability is high, since many docs are collected and gap approximation
+     * uses more expensive math
+     */
+    private static class RandomSamplingDocIdSetIterator extends DocIdSetIterator {
+        int doc = -1;
+        final int maxDoc;
+        final Random rng;
+        final double p;
+
+        RandomSamplingDocIdSetIterator(int maxDoc, Random rng, double p) {
+            this.maxDoc = maxDoc;
+            this.rng = rng;
+            this.p = p;
+        }
+
+        @Override
+        public int docID() {
+            return doc;
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return advance(doc + 1);
+        }
+
+        @Override
+        public int advance(int target) throws IOException {
+            doc = target;
+            while (doc < maxDoc) {
+                if (rng.nextDouble() <= p) {
+                    return doc;
+                }
+                doc = doc + 1;
+            }
+            return NO_MORE_DOCS;
```

Comment:
And can the doc be the default, but the value be optionally pulled from a field? See https://github.com/elastic/elasticsearch/issues/25240#issuecomment-310501468. Otherwise there is no way to have real consistency with a seed (and IMO seed should only be allowed when a field is specified).